src/persist/src/storage.rs

```diff
@@ -330,6 +330,167 @@ impl From<(&str, &str)> for LockInfo {
+
+    /// List all of the keys in the map that currently have a value.
+    fn list_keys(&self) -> Result<Vec<String>, Error>;
+
+    /// Conditionally insert/update a key-value pair into the map, if the existing
+    /// value matches expected.
+    ///
+    /// Return true if the cas executed correctly, and false if the current value != expected.
+    ///
+    /// TODO: could also have the return value be the current value here. Doing a simpler thing
+    /// for prototyping.
+    fn cas(
+        &mut self,
+        key: &str,
+        value: Option<Vec<u8>>,
+        expected: Option<Vec<u8>>,
+    ) -> Result<bool, Error>;
+}
+
+/// Abstraction over a key value store that supports atomic inserts, and scans
+/// from a monotonically increasing index
+///
+/// "real world" realizations of this api could be kafka / redpanda.
+pub trait AtomicQueue {
+    /// Append data to the queue, and return the index it was appended at.
+    ///
+    /// Indexes increase monotonically.
+    fn append(&mut self, data: Vec<u8>) -> Result<u64, Error>;
+
+    /// Read all data stored at indexes >= index.
+    fn scan(&self, index: u64) -> Result<Vec<(u64, Vec<u8>)>, Error>;
+
+    /// Delete all data stored at indexes < index
+    fn truncate(&mut self, index: u64) -> Result<(), Error>;
+}
+
+/// WIP
+pub trait Consensus {
+    /// Return the most recently stored data
+    fn head(&self) -> Result<Option<(SeqNo, Vec<u8>)>, Error>;
+
+    /// Atomically insert data at seqno, iff the current HEAD of the data is at seqno - 1.
+    ///
+    /// Return true if the insert is successful (seqno - 1 was the previous version and this
+    /// call to cas wrote at seqno), and false otherwise.
+    fn cas(&mut self, seqno: SeqNo, data: Vec<u8>) -> Result<bool, Error>;
+
+    /// Read all data inserted with sequence numbers >= seqno
+    fn scan(&self, seqno: SeqNo) -> Result<Vec<(SeqNo, Vec<u8>)>, Error>;
+
+    /// Delete all data stored at sequence numbers < seqno
+    fn truncate(&mut self, seqno: SeqNo) -> Result<(), Error>;
+}
+
+const HEAD_KEY: &'static str = "HEAD";
+
+impl<T: CasKeyVal> Consensus for T {
+    fn head(&self) -> Result<Option<(SeqNo, Vec<u8>)>, Error> {
+        let head = self.get(HEAD_KEY)?;
+
+        let head = match head {
+            Some(head) => head,
+            None => return Ok(None),
+        };
+
+        let key = std::str::from_utf8(&head).expect("wip");
+        let seqno = key.parse::<u64>().expect("wip");
+
+        let value = self.get(key)?.expect("wip");
+
+        Ok(Some((SeqNo(seqno), value)))
+    }
+
+    fn cas(&mut self, seqno: SeqNo, data: Vec<u8>) -> Result<bool, Error> {
+        // First, write out the data
+        let key = seqno.0.to_string();
+
+        let write_success = self.cas(&key, Some(data.clone()), None)?;
+
+        // Someone already wrote to this seqno - exit.
+        if !write_success {
+            return Ok(false);
+        }
+
+        // Attempt to repoint HEAD from seqno - 1 to seqno
```

Comment:
I feel relatively strongly that this need to be one write, so concretely I object to maintaining a separate HEAD

my latest thinking on this is that Consensus should look something like `get` which returns some recently written value (ideally latest, but y'know, distributed systems) and `cas` which updates it. then, whatever use led me to originally surface previous versions of the Consensus value in the prototype is something we should push into explicitly modeling in `State`

Reply:
yeah, agreed! I'm pretty sure we're saying the same things here (though you're saying it much more clearly and in more detail than I am)

> I also feel like its not like building an implementation of the Consensus API i laid out is impossible with one write. I think it just means we need different / stronger primitives. Concretely, I think if we had access to a max_key function or something like that (think SELECT max(seqno) FROM table or something like that) then we could use that as HEAD and avoid having to store/repoint a HEAD pointer

yeah, I'm also curious about exploring this direction. my big concern with this idea is the other comment I left about these keys accumulating an unbounded number of deletion tombstones in some way

Reply:
I forgot to ask earlier:

> I feel relatively strongly that this need to be one write, so concretely I object to maintaining a separate HEAD

Whats driving this feeling (is it more like "want to minimize costs" or is it more like "want to keep consensus simpler" or is it something else)?

Reply:
latency :D!

Reply:
gotcha. do you have a rough order of magnitude sense of what latency you are targeting? concretely I was imagining targeting something with a p25 of 10 ms and a p75 latency of 20 - 25 ms -- does that sound roughly right to you?

Reply:
those are both pretty fast given the s3 latencies I posed in the slack channel the other day. I think it's important to look both at actual end-to-end latencies as well as a theoretical model of them. in this case, I think the hot path for any common operation should be either of: `1 write to s3 + 1 write to consensus` OR `1 write to consensus`